service/collector.go

```diff
@@ -163,13 +163,49 @@ LOOP:
 	return col.shutdown(ctx)
 }
 
+func printFactories(factories component.Factories) {
+	ids := make([]string, len(factories.Extensions))
+	i := 0
+	for id := range factories.Extensions {
+		ids[i] = string(id)
+		i++
+	}
+	fmt.Printf("Available extensions are: %v\n", ids)
+
+	ids = make([]string, len(factories.Receivers))
+	i = 0
+	for id := range factories.Receivers {
+		ids[i] = string(id)
+		i++
+	}
+	fmt.Printf("Available receivers are: %v\n", ids)
+
+	ids = make([]string, len(factories.Processors))
+	i = 0
+	for id := range factories.Processors {
+		ids[i] = string(id)
+		i++
+	}
+	fmt.Printf("Available processors are: %v\n", ids)
+
+	ids = make([]string, len(factories.Exporters))
+	i = 0
+	for id := range factories.Exporters {
+		ids[i] = string(id)
+		i++
+	}
+	fmt.Printf("Available exporters are: %v\n", ids)
+}
+
 // setupConfigurationComponents loads the config and starts the components. If all the steps succeeds it
 // sets the col.service with the service currently running.
 func (col *Collector) setupConfigurationComponents(ctx context.Context) error {
 	col.setCollectorState(Starting)
 
 	cfg, err := col.set.ConfigProvider.Get(ctx, col.set.Factories)
 	if err != nil {
+		printFactories(col.set.Factories)
```

Comment:
Do we print this on every error? This may be annoying for errors that are not because the component name is wrong. Can we ensure we only print the factories list for the specific error?

Reply:
@tigrannajaryan This code prints the factories for every error.

I considered three implementations:
- Always printing or logging the factories at startup
- Printing only if there is some error
- Printing the factories only if there is this specific problem

It would be straightforward to add a parameter to `errorUnknownType()` so that the error would be very specific: 

```
collector server run finished with error: cannot unmarshal the configuration: unknown exporters type "logging-typo" for "logging-typo (acceptable types are otlp, otlphttp, logging)"
```

Shall I redo the PR to put the valid choices directly into the error?
Or shall I keep the current output, but suppress it unless the problem was generated by errorUnknownType()?

Reply:
@tigrannajaryan I have a different PR which only outputs on error, see https://github.com/open-telemetry/opentelemetry-collector/pull/4684